File: Kinect Client WebGL/Assets/Scripts/KinectManager.js
const KINECT_XBOX360_JOINT_COUNT = 20;
const KINECT_XBOXONE_JOINT_COUNT = 25;

class UserPosition {
    get position() {
        return this._position;
    }

    get rotation() {
        return this._rotation;
    }

    set position(value) {
        this._position.x = value.x;
        this._position.y = value.y;
        this._position.z = value.z;
    }

    set rotation(value) {
        this._rotation.x = value.x;
        this._rotation.y = value.y;
        this._rotation.z = value.z;
        this._rotation.w = value.w;
    }

    constructor() {
        this._position = new BABYLON.Vector3(0, 0, 0);
        this._rotation = new BABYLON.Quaternion(0, 0, 0, 1);
    }
}

class KinectManager {
    get initialized() {
        return this._initialized;
    }

    get bodies() {
        return this._bodies;
    }

    get kinectVersion() {
        if (this._bodyCount > 0) {
            return this._bodies[0].Version;
        }

        return -1;
    }

    get isKienctV1() {
        if (this._bodyCount > 0) {
            return this._bodies[0].Version == 1;
        }

        return false;
    }

    get isKinectV2() {
        if (this._bodyCount > 0) {
            return this._bodies[0].Version == 2;
        }

        return false;
    }

    get isConnected() {
        return this._bodyCount > 0;
    }

    constructor(ip, port) {
        this._bodies = null;
        this._bodyCount = 0;
        this._usersTransform = [];
        this._spinBaseJoint = null;

        this._webSocket = new WebSocket("ws://" + ip + ":" + port);
        this._webSocket.onopen = (e) => console.log("[KinectManager] Connected to the Kinect Server.");
        this._webSocket.onerror = (e) => console.log("[KinectManager] An error was detected.");
        this._webSocket.onmessage = (e) => this._onMessage(e);
    }

    dispose() {
        this._webSocket.close();
    }

    getJoint(type, userIndex) {
        if (this._bodyCount > 0) {
            if (this.isKinectV2 || type < KINECT_XBOX360_JOINT_COUNT) {
                return this._bodies[userIndex].Joints[type];
            }
        }

        return new KinectJoint();
    }

    getBody(userIndex) {
        if (this._bodyCount > userIndex) {
            return this._bodies[userIndex];
        }

        return null;
    }

    getJointPosition(type, userIndex) {
        return this.getJoint(type, userIndex).Position;
    }

    getJointOrientation(type, userIndex) {
        return this.getJoint(type, userIndex).Rotation;
    }

    getUserPosition(userIndex) {
        if (this._bodyCount > 0) {
            return this._usersTransform[userIndex].position;
        }

        return BABYLON.Vector3.Zero;
    }

    getUserRotation(userIndex) {
        if (this._bodyCount > 0) {
            return this._usersTransform[userIndex].rotation;
        }

        return new BABYLON.Quaternion(0, 0, 0, 1);
    }

    getHandState(handLeft, userIndex) {
        if (this._bodyCount > 0) {
            return handLeft ? this._bodies[userIndex].HandLeftState : this._bodies[userIndex].HandRightState;
        }

        return 0;
    }

    getHandConfidence(handLeft, userIndex) {
        if (this._bodyCount > 0) {
            return handLeft ? this._bodies[userIndex].HandLeftConfidence : this._bodies[userIndex].HandRightConfidence;
        }

        return 0;
    }

    getParentJoint(type, userIndex) {
        var parent = null;

        switch (type) {
            case JointType.SpineMid:
            case JointType.HipLeft:
            case JointType.HipRight:
            case JointType.KneeLeft:
            case JointType.KneeRight:
            case JointType.AnkleLeft:
            case JointType.AnkleRight:
            case JointType.FootLeft:
            case JointType.FootRight:
                parent = this.getJoint(JointType.SpineBase, userIndex);
                break;

            case JointType.SpineShoulder:
                parent = this.getJoint(JointType.SpineMid, userIndex);
                break;

            case JointType.Head:
            case JointType.Neck:
            case JointType.ShoulderLeft:
            case JointType.ShoulderRight:
            case JointType.ElbowLeft:
            case JointType.ElbowRight:
            case JointType.WristLeft:
            case JointType.WristRight:
                parent = this.getJoint(JointType.SpineShoulder, userIndex);
                break;

            case JointType.HandRight:
            case JointType.ThumbRight:
                parent = this.getJoint(JointType.WristRight, userIndex);
                break;

            case JointType.HandTipRight:
                parent = this.getJoint(JointType.HandRight, userIndex);
                break;

            case JointType.HandLeft:
            case JointType.ThumbLeft:
                parent = this.getJoint(JointType.WristLeft, userIndex);
                break;

            case JointType.HandTipLeft:
                parent = this.getJoint(JointType.HandLeft, userIndex);
                break;
        }

        return parent;
    }

    _onMessage(event) {
        this._bodies = JSON.parse(event.Data);

        if (this._bodies !== null) {
            this._bodyCount = this._bodies.length;

            for (var i = 0; i < this._bodyCount; i++) {
                this._spinBaseJoint = this.getJoint(JointType.SpineBase, i);
                this._usersTransform[i].position = this._spinBaseJoint.Position;
                this._usersTransform[i].rotation = this._spinBaseJoint.Rotation;
            }

            if (!this._initialized) {
                this._initialized = true;
            }
        } else {
            this._bodyCount = 0;
        }
    }
}

File: Kinect Client Unity/Assets/Sources/KinectManager.cs
using UnityEngine;
using WebSocketSharp;
using Newtonsoft.Json;

namespace Demonixis.Kinect
{
    public struct UserPosition
    {
        public Vector3 Position { get; set; }
        public Quaternion Rotation { get; set; }
    }

    public sealed class KinectManager : MonoBehaviour
    {
        #region Constants and Static Fields

        public const int KinectXbox360JointCount = 20;
        public const int KinectXboxOneJointCount = 25;
        private static KinectManager _instance = null;

        #endregion

        #region Private Variables

        private WebSocket _webSocket = null;
        private Body[] _bodies;
        private int _bodyCount = 0;
        private UserPosition[] _usersTransform;
        private Joint _dummyJoint = new Joint();
        private Joint _spinBaseJoint = new Joint();

        #endregion

        #region Editor Fields

        [Header("Server Settings")]
        [SerializeField]
        private string _ip = "127.0.0.1";
        [SerializeField]
        private string _port = "8181";
        [SerializeField]
        private bool _showLog = true;
        [SerializeField]
        private bool _autoConnect = true;

        #endregion

        #region Properties

        public static KinectManager Instance
        {
            get
            {
                if (_instance == null)
                    _instance = FindObjectOfType<KinectManager>();

                return _instance;
            }
        }

        public bool IsInitialized
        {
            get; private set;
        }

        public Body[] Bodies
        {
            get { return _bodies; }
        }

        public int KinectVersion
        {
            get { return _bodyCount > 0 ? _bodies[0].Version : 0; }
        }

        public bool IsKinectV1
        {
            get { return _bodyCount > 0 ? _bodies[0].Version == 1 : false; }
        }

        public bool IsKinectV2
        {
            get { return _bodyCount > 0 ? _bodies[0].Version == 2 : false; }
        }

        public bool IsConnected
        {
            get { return _bodyCount > 0; }
        }

        #endregion

        #region Lifecycle

        void Awake()
        {
            if (_instance != null)
            {
                Debug.Log("[KinectManager] Multiple instances is not supported.");
                Debug.Log(string.Format("[KinectManager] The object attached to {0} will be destroyed", gameObject.name));
                Destroy(this);
            }

            _instance = this;

            if (_autoConnect)
                Connect(_ip, _port);
        }

        void OnDestroy()
        {
            if (_webSocket != null)
            {
                _webSocket.OnMessage -= OnMessage;
                _webSocket.Close();
            }
        }

        public void Connect()
        {
            Connect(_ip, _port);
        }

        public void Connect(string ip, string port)
        {
            if (_webSocket != null)
            {
                _webSocket.OnMessage -= OnMessage;
                _webSocket.Close();
                _webSocket = null;
            }

            _webSocket = new WebSocket(string.Format("ws://{0}:{1}", _ip, _port));

            if (_showLog)
                _webSocket.OnOpen += (s, e) => Debug.Log("[KinectManager] Connected to the Kinect Server.");

            if (_showLog)
                _webSocket.OnError += (s, e) => Debug.LogFormat("[KinectManager] An Error was detected.\n{0}", e.Message);

            _webSocket.Connect();
            _webSocket.OnMessage += OnMessage;
        }

        #endregion

        #region Methods to get data on a body

        public Joint GetJoint(JointType type, int userIndex)
        {
            if (_bodyCount > userIndex)
            {
                var index = (int)type;
                if (IsKinectV2 || index < KinectXbox360JointCount)
                    return _bodies[userIndex].Joints[index];
            }

            return _dummyJoint;
        }

        public Body GetBody(int userIndex)
        {
            if (_bodyCount > userIndex)
                return _bodies[userIndex];

            return null;
        }

        public Vector3 GetJointPosition(JointType type, int userIndex)
        {
            return GetJoint(type, userIndex).Position;
        }

        public Quaternion GetJointOrientation(JointType type, int userIndex)
        {
            return GetJoint(type, userIndex).Rotation;
        }

        public Vector3 GetUserPosition(int userIndex)
        {
            if (_bodyCount > userIndex)
                return _usersTransform[userIndex].Position;

            return Vector3.zero;
        }

        public Quaternion GetUserRotation(int userIndex)
        {
            if (_bodyCount > userIndex)
                return _usersTransform[userIndex].Rotation;

            return Quaternion.identity;
        }

        public HandState GetHandState(bool handLeft, int userIndex)
        {
            if (_bodyCount > userIndex)
                return handLeft ? _bodies[userIndex].HandLeftState : _bodies[userIndex].HandRightState;
            else
                return HandState.Unknown;
        }

        public TrackingConfidence GetHandConfidence(bool handLeft, int userIndex)
        {
            if (_bodyCount > userIndex)
                return handLeft ? _bodies[userIndex].HandLeftConfidence : _bodies[userIndex].HandRightConfidence;
            else
                return TrackingConfidence.Low;
        }

        public Joint GetParentJoint(JointType type, int userIndex)
        {
            Joint parent = null;

            switch (type)
            {
                case JointType.SpineMid:
                case JointType.HipLeft:
                case JointType.HipRight:
                case JointType.KneeLeft:
                case JointType.KneeRight:
                case JointType.AnkleLeft:
                case JointType.AnkleRight:
                case JointType.FootLeft:
                case JointType.FootRight:
                    parent = GetJoint(JointType.SpineBase, userIndex);
                    break;

                case JointType.SpineShoulder:
                    parent = GetJoint(JointType.SpineMid, userIndex);
                    break;

                case JointType.Head:
                case JointType.Neck:
                case JointType.ShoulderLeft:
                case JointType.ShoulderRight:
                case JointType.ElbowLeft:
                case JointType.ElbowRight:
                case JointType.WristLeft:
                case JointType.WristRight:
                    parent = GetJoint(JointType.SpineShoulder, userIndex);
                    break;

                case JointType.HandRight:
                case JointType.ThumbRight:
                    parent = GetJoint(JointType.WristRight, userIndex);
                    break;

                case JointType.HandTipRight:
                    parent = GetJoint(JointType.HandRight, userIndex);
                    break;

                case JointType.HandLeft:
                case JointType.ThumbLeft:
                    parent = GetJoint(JointType.WristLeft, userIndex);
                    break;

                case JointType.HandTipLeft:
                    parent = GetJoint(JointType.HandLeft, userIndex);
                    break;
            }

            return parent;
        }

        #endregion

        #region WebSocket Handler

        private void OnMessage(object sender, MessageEventArgs e)
        {
            _bodies = JsonConvert.DeserializeObject<Body[]>(e.Data);

            if (_bodies != null)
            {
                _bodyCount = _bodies.Length;

                if (_usersTransform == null || _usersTransform.Length != _bodyCount)
                    _usersTransform = new UserPosition[_bodyCount];

                for (int i = 0; i < _bodyCount; i++)
                {
                    _spinBaseJoint = GetJoint(JointType.SpineBase, i);
                    _usersTransform[i].Position = _spinBaseJoint.Position;
                    _usersTransform[i].Rotation = _bodies[i].RootOrientation;
                }

                if (!IsInitialized)
                    IsInitialized = true;
            }
            else
                _bodyCount = 0;
        }

        #endregion
    }
}